tools/code_samples/code_samples_usage.php

```diff
@@ -0,0 +1,213 @@
+        $includingFileList = getIncludingFileList($codeSampleFile);
+        foreach ($includingFileList as $includingFile) {
+            $blocks = getInclusionBlocks($includingFile, $codeSampleFile);
+            displayBlocks($blocks, $includingFile);
+        }
+    }
+} else {
+    $includingFileList = getIncludingFileList();
+    foreach ($includingFileList as $includingFile) {
+        $blocks = getInclusionBlocks($includingFile);
+        displayBlocks($blocks, $includingFile);
+    }
+}
+
+function displayBlocks(array $docFileBlocks, string $docFilePath = null, $lineOffset = 0): void
+{
+    if (!$docFilePath) {
+        $docFilePath = '';
+    }
+    foreach ($docFileBlocks as $block) {
+        $prefixedBlockLines = [];
+        foreach ($block as $blockLineIndex => $blockLine) {
+            $lineNumber = $blockLineIndex + $lineOffset;
+            $prefixedBlockLines[$blockLineIndex] = "$docFilePath@$lineNumber:$blockLine";
+        }
+        echo implode(PHP_EOL, $prefixedBlockLines) . PHP_EOL . PHP_EOL;
+
+        $prefixedBlockContentLines = [];
+        try {
+            $blockContents = getBlockContents($block);
+            foreach ($blockContents['contents'] as $contentLineNumber => $contentLine) {
+                $prefixedBlockContentLines[] = str_pad($contentLineNumber, 3, 0, STR_PAD_LEFT) . (in_array($contentLineNumber, $blockContents['highlights']) ? '⫸' : '⫶') . $contentLine;
+            }
+            echo implode(PHP_EOL, $prefixedBlockContentLines) . PHP_EOL . PHP_EOL;
+        } catch (Exception $exception) {
+            echo $exception->getMessage() . PHP_EOL . PHP_EOL;
+        }
+    }
+}
+
+
+// FRAMEWORK
+
+/**
+ * @param string $targetBranch Branch to compare to.
+ * @param string $sourceBranch Branch to compare from. By default, the current branch HEAD
+ * @return array<int, string> List of file paths from code_samples/ directory.
+ */
+function getModifiedCodeSampleFileList(string $targetBranch = 'origin/master', string $sourceBranch = 'HEAD'): array
+{
+    $command = "git diff --name-only $sourceBranch..$targetBranch -- code_samples";
+    exec($command, $rawModifiedCodeSampleList, $commandResultCode);
+    if (0 === $commandResultCode) {
+        return $rawModifiedCodeSampleList;
+    }
+    throw new RuntimeException("The following Git command failed: $command");
+}
+
+/**
+ * @param string $codeSampleFilePath The path from project root to an included file.
+ *                                  If given, files including this one are returned.
+ *                                  If null, all files including a file are returned.
+ * @return array<int, string> List of file paths from docs/ directory.
+ */
+function getIncludingFileList(string $codeSampleFilePath = null): array
+{
+    $pattern = null === $codeSampleFilePath ? '= include_file' : $codeSampleFilePath;
+    $command = "grep '$pattern' -Rl docs | sort";
+    exec($command, $rawIncludingFileList, $commandResultCode);
```

Comment:
I'm not sure there is an issue here - but just to be sure that there's no chance of command injection with some crazy filenames - could you use the Process component (https://symfony.com/doc/current/components/process.html)?

I'd also limit the scope of the token to the minimum (https://docs.github.com/en/actions/writing-workflows/choosing-what-your-workflow-does/controlling-permissions-for-github_token#defining-access-for-the-github_token-permissions) for this job, just for safety

Reply:
For now I used [`escapeshellarg`](https://www.php.net/manual/function.escapeshellarg.php) around the file path in ebc14f5e2b52b8924d0b9e763ad66478dd00f883.